<h2>EBA3400 Exercise Set 8: Introduction to Pandas</h2>

<h3>8.1 Melbourne housing data</h3>

In this exercise we will investigate the Melbourne housing market. The dataset we will work with consists of almost 15k residential houses and several variables describing these houses. Perform the following tasks

- A Import the csv file melbourne.csv using pandas
- B Investigate the data. How many observations are there? How many variables are stored for each house? 
- C Create a subset of the data consisting of the variables Suburb, Price, Rooms, Type, Bathroom, Car, Landsize, BuildingArea and YearBuilt.
- D Show the summary statistics for these variables
- E What is the average price? What is the maximum number of rooms? What is the average number of baths? Are there bathrooms in all houses? In which year was the newest house built?
- F What is the average price for all houses with at least four rooms and two bathrooms in the Newport area? How many houses are in this area?

In [10]:
import pandas as pd 

<h4>Part A</h4>

In [11]:
df = pd.read_csv('C:/Users/power/OneDrive/Documents/Programming/Data/melbourne.csv')

<h4>Part B</h4>

In [3]:
df.head(4)

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  

[4 rows x 21 columns]

In [4]:
print('Number of observations:', len(df))
print('Number or varibles:', len(df.columns))

Number of observations: 13580
Number or varibles: 21


<h4>Part C</h4>

In [5]:
#subset of the data
index_col = ['Suburb', 'Price', 'Rooms', 'Type', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'YearBuilt']
#creating the subset
df_sub = df[index_col]
#viewing the subset 
df_sub.head(4)

Suburb      Price  Rooms Type  Bathroom  Car  Landsize  BuildingArea  \
0  Abbotsford  1480000.0      2    h       1.0  1.0     202.0           NaN   
1  Abbotsford  1035000.0      2    h       1.0  0.0     156.0          79.0   
2  Abbotsford  1465000.0      3    h       2.0  0.0     134.0         150.0   
3  Abbotsford   850000.0      3    h       2.0  1.0      94.0           NaN   

   YearBuilt  
0        NaN  
1     1900.0  
2     1900.0  
3        NaN

<h4>Part D</h4>

In [6]:
#summary statistics
df_sub.describe()

Price         Rooms      Bathroom           Car       Landsize  \
count  1.358000e+04  13580.000000  13580.000000  13518.000000   13580.000000   
mean   1.075684e+06      2.937997      1.534242      1.610075     558.416127   
std    6.393107e+05      0.955748      0.691712      0.962634    3990.669241   
min    8.500000e+04      1.000000      0.000000      0.000000       0.000000   
25%    6.500000e+05      2.000000      1.000000      1.000000     177.000000   
50%    9.030000e+05      3.000000      1.000000      2.000000     440.000000   
75%    1.330000e+06      3.000000      2.000000      2.000000     651.000000   
max    9.000000e+06     10.000000      8.000000     10.000000  433014.000000   

       BuildingArea    YearBuilt  
count   7130.000000  8205.000000  
mean     151.967650  1964.684217  
std      541.014538    37.273762  
min        0.000000  1196.000000  
25%       93.000000  1940.000000  
50%      126.000000  1970.000000  
75%      174.000000  1999.000000  
max    44515.000000  2018.000000

<h4>Part E</h4>

- Average price: 1.076
- Max no. of rooms: 10
- Average no. of baths: 1.534
- Bathrooms in all houses: some have 0 (bathroom - min)
- Year the newest house was built: 2018

<h4>Part F</h4>

In [7]:
# Getting the id of the relevant rows 
idx = (df.Suburb == 'Newport') & (df.Rooms >= 4) & (df.Bathroom >= 2)
#using the describe function
df_sub[index_col].describe()


Price         Rooms      Bathroom           Car       Landsize  \
count  1.358000e+04  13580.000000  13580.000000  13518.000000   13580.000000   
mean   1.075684e+06      2.937997      1.534242      1.610075     558.416127   
std    6.393107e+05      0.955748      0.691712      0.962634    3990.669241   
min    8.500000e+04      1.000000      0.000000      0.000000       0.000000   
25%    6.500000e+05      2.000000      1.000000      1.000000     177.000000   
50%    9.030000e+05      3.000000      1.000000      2.000000     440.000000   
75%    1.330000e+06      3.000000      2.000000      2.000000     651.000000   
max    9.000000e+06     10.000000      8.000000     10.000000  433014.000000   

       BuildingArea    YearBuilt  
count   7130.000000  8205.000000  
mean     151.967650  1964.684217  
std      541.014538    37.273762  
min        0.000000  1196.000000  
25%       93.000000  1940.000000  
50%      126.000000  1970.000000  
75%      174.000000  1999.000000  
max    44515.000000  2018.000000

<h3>8.2 Cleaning a biking dataset </h3>

In this exercise we will tidy up a dataset consisting of cyclist data from Montrèal in 2018. You are asked to do the following:

- A Import the csv file bikes.csv. Do you get an error message? The file is in French and uses some characters that needs to be accounted for before import. This can be done by specifying encoding='latin1' inside the read_csv method.
- B Look at the three first rows? What is wrong here?
- C To clean the data, do the following inside the read_csv method:
    - Change the column separator to a ; (using sep)
    - Parse the dates in the 'Date' column (use parse_dates)
    - Tell it that our dates have the day first instead of the month first (use dayfirst=True)
    - Set the index to be the 'Date' column (use index_col)

<h4>Part A</h4>

In [15]:
# Import bike data with latin characters
df = pd.read_csv('C:/Users/power/OneDrive/Documents/Programming/Data/bikes.csv',encoding='latin1')

<h4>Part B</h4>

In [16]:
# Look at the data
df.head(3)

Date;Berri 1;Brébeuf (données non disponibles);Côte-Sainte-Catherine;Maisonneuve 1;Maisonneuve 2;du Parc;Pierre-Dupuy;Rachel1;St-Urbain (données non disponibles)
0                   01/01/2012;35;;0;38;51;26;10;16;                                                                                                               
1                   02/01/2012;83;;1;68;153;53;6;43;                                                                                                               
2                 03/01/2012;135;;2;104;248;89;3;58;

<h4>Part C</h4>

In [18]:
# Import again with correct options
df = pd.read_csv('C:/Users/power/OneDrive/Documents/Programming/Data/bikes.csv',encoding='latin1',sep=';',parse_dates=['Date'],dayfirst=True,index_col='Date')

In [19]:
# View imported data
df.head(5)

Berri 1  Brébeuf (données non disponibles)  Côte-Sainte-Catherine  \
Date                                                                            
2012-01-01       35                                NaN                      0   
2012-01-02       83                                NaN                      1   
2012-01-03      135                                NaN                      2   
2012-01-04      144                                NaN                      1   
2012-01-05      197                                NaN                      2   

            Maisonneuve 1  Maisonneuve 2  du Parc  Pierre-Dupuy  Rachel1  \
Date                                                                       
2012-01-01             38             51       26            10       16   
2012-01-02             68            153       53             6       43   
2012-01-03            104            248       89             3       58   
2012-01-04            116            318      111             8       61   
2012-01-05            124            330       97            13       95   

            St-Urbain (données non disponibles)  
Date                                             
2012-01-01                                  NaN  
2012-01-02                                  NaN  
2012-01-03                                  NaN  
2012-01-04                                  NaN  
2012-01-05                                  NaN

<h3>8.3 Complaints in New York City </h3>

In the US, 311 is a non-emergency phone number that people can call in many cities to find information about services, make complaints, or report problems like graffiti or road damage. This dataset is updated in close to real time, currently counting above 21 million complaints and 41 variables. We can get access to the data from the NYC Open Data website, but for the purpose of this exercise we only use a subset of the data available in the file complaints.csv.

- A Import the data
- B How many observations and variables are there in the data?
- C View the first five records
- D View rows 2,3,4 and 9 of the column Complaint Type 
- E Create a subset of the data, consisting of only Complaint Type and Borough
- F How many complaints are recorded for each of the NYC regions? Use the method .value_counts()
- G What are the most common complaint types?
- H Create a top 10 list of the most common complaint types and a top 3 list for the regions with the most complaints

<h4>Part A</h4>

In [21]:
# Import data
df = pd.read_csv('C:/Users/power/OneDrive/Documents/Programming/Data/complaints.csv')

C:\Users\power\AppData\Local\Temp\ipykernel_12840\1073461392.py:2: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('C:/Users/power/OneDrive/Documents/Programming/Data/complaints.csv')


<h4>Part B</h4>

In [22]:
# Print observations and variables
print('Number of observations:',len(df))
print('Number of variables:',len(df.columns))

Number of observations: 111069
Number of variables: 52


<h4>Part C</h4>

In [23]:
# View first five
df.head(5)

Unique Key            Created Date             Closed Date Agency  \
0    26589651  10/31/2013 02:08:41 AM                     NaN   NYPD   
1    26593698  10/31/2013 02:01:04 AM                     NaN   NYPD   
2    26594139  10/31/2013 02:00:24 AM  10/31/2013 02:40:32 AM   NYPD   
3    26595721  10/31/2013 01:56:23 AM  10/31/2013 02:21:48 AM   NYPD   
4    26590930  10/31/2013 01:53:44 AM                     NaN  DOHMH   

                               Agency Name           Complaint Type  \
0          New York City Police Department  Noise - Street/Sidewalk   
1          New York City Police Department          Illegal Parking   
2          New York City Police Department       Noise - Commercial   
3          New York City Police Department          Noise - Vehicle   
4  Department of Health and Mental Hygiene                   Rodent   

                     Descriptor        Location Type Incident Zip  \
0                  Loud Talking      Street/Sidewalk      11432.0   
1  Commercial Overnight Parking      Street/Sidewalk      11378.0   
2              Loud Music/Party  Club/Bar/Restaurant      10032.0   
3                Car/Truck Horn      Street/Sidewalk      10023.0   
4  Condition Attracting Rodents           Vacant Lot      10027.0   

   Incident Address  ... Bridge Highway Name Bridge Highway Direction  \
0  90-03 169 STREET  ...                 NaN                      NaN   
1         58 AVENUE  ...                 NaN                      NaN   
2     4060 BROADWAY  ...                 NaN                      NaN   
3    WEST 72 STREET  ...                 NaN                      NaN   
4   WEST 124 STREET  ...                 NaN                      NaN   

  Road Ramp Bridge Highway Segment Garage Lot Name Ferry Direction  \
0       NaN                    NaN             NaN             NaN   
1       NaN                    NaN             NaN             NaN   
2       NaN                    NaN             NaN             NaN   
3       NaN                    NaN             NaN             NaN   
4       NaN                    NaN             NaN             NaN   

  Ferry Terminal Name   Latitude  Longitude  \
0                 NaN  40.708275 -73.791604   
1                 NaN  40.721041 -73.909453   
2                 NaN  40.843330 -73.939144   
3                 NaN  40.778009 -73.980213   
4                 NaN  40.807691 -73.947387   

                                   Location  
0   (40.70827532593202, -73.79160395779721)  
1  (40.721040535628305, -73.90945306791765)  
2   (40.84332975466513, -73.93914371913482)  
3    (40.7780087446372, -73.98021349023975)  
4   (40.80769092704951, -73.94738703491433)  

[5 rows x 52 columns]

<h4>Part D</h4>

In [24]:
# View rows 2,3,4 and 9 of the column Complaint Type 
df['Complaint Type'][[1,2,3,8]]

1       Illegal Parking
2    Noise - Commercial
3       Noise - Vehicle
8    Noise - Commercial
Name: Complaint Type, dtype: object

<h4>Part E</h4>

In [25]:
# Subset
df_sub = df[['Complaint Type','Borough']]
df_sub.head(5)

Complaint Type    Borough
0  Noise - Street/Sidewalk     QUEENS
1          Illegal Parking     QUEENS
2       Noise - Commercial  MANHATTAN
3          Noise - Vehicle  MANHATTAN
4                   Rodent  MANHATTAN

<h4>Part F</h4>

In [26]:
# Number of complaints by region
complaints_region = df_sub['Borough'].value_counts()
print(complaints_region)

BROOKLYN         32890
MANHATTAN        24288
QUEENS           22281
BRONX            19686
Unspecified       7107
STATEN ISLAND     4817
Name: Borough, dtype: int64


<h4>Part G</h4>

In [27]:
# Most common complaint types
complaints_type = df_sub['Complaint Type'].value_counts()
print(complaints_type)

HEATING                           14200
GENERAL CONSTRUCTION               7471
Street Light Condition             7117
DOF Literature Request             5797
PLUMBING                           5373
                                  ...  
Municipal Parking Facility            1
Tunnel Condition                      1
DHS Income Savings Requirement        1
Stalled Sites                         1
X-Ray Machine/Equipment               1
Name: Complaint Type, Length: 165, dtype: int64


<h4>Part H</h4>

In [28]:
# Top 10 list of complaint types
top10_types = complaints_type[:10]

# Top 3 list of regions with most complaints
top3_regions = complaints_region[:3]

In [29]:
# Print results
print(top10_types)

HEATING                   14200
GENERAL CONSTRUCTION       7471
Street Light Condition     7117
DOF Literature Request     5797
PLUMBING                   5373
PAINT - PLASTER            5149
Blocked Driveway           4590
NONCONST                   3998
Street Condition           3473
Illegal Parking            3343
Name: Complaint Type, dtype: int64


In [30]:
# Print results
print(top3_regions)

BROOKLYN     32890
MANHATTAN    24288
QUEENS       22281
Name: Borough, dtype: int64
